lb_content_resolver/subsonic.py

```diff
@@ -0,0 +1,132 @@
+                    album_mbid = album_info["albumInfo"]["musicBrainzId"]
+                except KeyError:
+                    print("subsonic album '%s' by '%s' has no MBID" % (album["album"], album["artist"]))
+                    continue
+
+                cursor.execute(
+                    """SELECT recording.id
+                                       , track_num
+                                       , COALESCE(disc_num, 1)
+                                    FROM recording
+                                   WHERE release_mbid = ?""", (album_mbid, ))
+
+                # create index on (track_num, disc_num)
+                release_tracks = {(row[1], row[2]): row[0] for row in cursor.fetchall()}
+
+                album_info = conn.getAlbum(id=album["id"])
+
+                if len(release_tracks) == 0:
+                    print("For album %s" % album_mbid)
+                    print("loaded %d of %d expected tracks from DB." % (len(release_tracks), len(album_info["album"]["song"])))
+
+                print("album '%s' by '%s'" % (album["album"], album["artist"]))
+                if "song" not in album_info["album"]:
+                    print("No songs returned")
+                else:
+                    for song in album_info["album"]["song"]:
+
+                        if (song["track"], song.get("discNumber", 1)) in release_tracks:
+                            recordings.append((release_tracks[(song["track"], song["discNumber"])], song["id"]))
+                        else:
+                            print("Song not matched: ", song["title"])
+                            continue
+
+            self.update_recordings(recordings)
+
+            print("fetched %d releases" % albums_this_batch)
+            if albums_this_batch < self.MAX_ALBUMS_PER_CALL:
+                break
+
+    def update_recordings(self, recordings):
+        """
+            Given a list of recording_subsonic records, update the DB.
+            Updates recording_id, subsonic_id, last_update
+        """
+
+        recordings = [(r[0], r[1], datetime.datetime.now()) for r in recordings]
+
+        cursor = db.connection().cursor()
+        cursor.executemany(
+            """INSERT INTO recording_subsonic (recording_id, subsonic_id, last_updated)
+                                    VALUES (?, ?, ?)
+                 ON CONFLICT DO UPDATE SET recording_id = excluded.recording_id
+                                         , subsonic_id = excluded.subsonic_id
+                                         , last_updated = excluded.last_updated""", recordings)
+
+    def upload_playlist(self, jspf):
+        """
+            Given a JSPF playlist, upload the playlist to the subsonic API.
+        """
+
+        conn = libsonic.Connection(config.SUBSONIC_HOST, config.SUBSONIC_USER, config.SUBSONIC_PASSWORD, config.SUBSONIC_PORT)
+
+        song_ids = [
+            track["extension"]["https://musicbrainz.org/doc/jspf#track"]["subsonic_identifier"][33:]
```

Comment:
I am not sure about the spec but IIRC this should go in additional_metadata like spotify ids. Best to check once where we do this in troi. Also, if it makes sense for this code to be re-usable it could go in troi.

Reply:
Good catch, fixed.